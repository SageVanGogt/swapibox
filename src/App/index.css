.App {
  display: grid;
  grid-template-columns: 200px 1fr 400px;
  grid-template-rows: 300px 1fr;
  grid-template-areas: "a b c" "e e f";
  justify-content: stretch;
  align-items: stretch; }

.nav-container {
  grid-area: f; }

.beer-nav {
  height: 400px; }

.cantina-image {
  position: absolute;
  height: 100%;
  z-index: -1; }
