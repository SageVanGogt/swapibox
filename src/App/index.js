import React, { Component } from 'react';
import Button from './../Button/index';
import Film from './../Film/index';
import CardContainer from './../CardContainer/index';
import fetchApiData from './../api';
import { getRandomFilm } from './../helper';
import './index.css';

class App extends Component {
  constructor() {
    super();

    this.state = {
      currentRandomFilm: {},
      currentSectionData: [],
      allData: {vehicles: [], planets: [], people: [], favorites: [], favoriteCount: 0}
    };
  }

  async componentDidMount() {
    const film = await this.fetchFilm();
    const currentRandomFilm = await getRandomFilm(film);
    this.setState({
      currentRandomFilm
    })
  }

  fetchFilm = async () => {
    const url = 'https://swapi.co/api/films/'
    const response = await fetch(url);
    const film = await response.json();
    return film;
  }

  handleClickEvent = async (event) => {
    const { name } = event.target;
    if(this.state.allData[name].length || name === 'favorites') {
      this.setStateToExistingData(name);
      return;
    }
   
    const currentSectionData = await fetchApiData(name); 
    const allData = {...this.state.allData, [name]: currentSectionData}
    await this.setState({
      currentSectionData, 
      allData
    })
  }

  setStateToExistingData = (name) => {
    const currentSectionData = this.state.allData[name];

    this.setState({
      currentSectionData
    })
  }

  handleFavorite = (event) => {
    const { value } = event.target;
    const currentChosen = this.state.currentSectionData.find(data => data.name === value)
    const favorites = this.state.allData.favorites;
    const favoriteExists = favorites.find(data => data.name === value);
    //includes filter out, if it does spread it in
    //refactor this area
    currentChosen.favorited = !currentChosen.favorited;

    if(favoriteExists) {
      const favoriteIndex = favorites.indexOf(favoriteExists);
      favorites.splice(favoriteIndex, 1);
    } else {
      favorites.push(currentChosen)
    }
    const favoriteCount = favorites.length
    const allData = {...this.state.allData, favorites, favoriteCount}
    
    this.setState({
      allData
    })
  }

  render() {
    return (
      <div className="App">
        <img 
          src="/assets/cantina.jpg" 
          alt="background image of cantina" 
          className="cantina-image"
        />
        <div className="nav-container">
          <nav>
            <Button 
              section={'people'} 
              handleClickEvent={this.handleClickEvent}
            />
            <Button 
              section={'planets'} 
              handleClickEvent={this.handleClickEvent}
            />
            <Button 
              section={'vehicles'} 
              handleClickEvent={this.handleClickEvent}
            />
            <Button 
              section={'favorites'} 
              handleClickEvent={this.handleClickEvent} 
              favoriteCount={this.state.allData.favoriteCount}
            />
          </nav>
          <img 
            src="./assets/beerTap.png" 
            alt="beertap navigation" 
            className="beer-nav"
          />
        </div>
        { 
          this.state.currentRandomFilm && 
          <Film 
            currentRandomFilm={this.state.currentRandomFilm}
          />
        }
        {
          this.state.currentSectionData &&
          <CardContainer
            currentSectionData={this.state.currentSectionData}
            handleFavorite={this.handleFavorite}
          />
        }
      </div>
    );
  }
}

export default App;
