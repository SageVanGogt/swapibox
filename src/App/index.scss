
.App {
  display: grid;
  grid-template-columns: 200px 1fr 400px;
  grid-template-rows: 300px 1fr;
  grid-template-areas: 
  "a f f"
  "e e e";
  justify-content: stretch;
  align-items: stretch;
  // background-image: url('/assets/cantina.jpg');
  // background-size: 120vw;
  // background-repeat: no-repeat;
}

.nav-container {
  grid-area: f;
  margin-left: 500px;
  margin-top: 50px;
}

.beer-nav {
  height: 600px;
}

.cantina-image {
  position: absolute;
  height: 100%;
  z-index: -1;
}

nav {
  display: flex;
}
