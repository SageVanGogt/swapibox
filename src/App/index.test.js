import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import App from './index';
import films from './../mockData/mockFilms';

describe('App component', () => {
  let mockFilms;

  beforeEach(() => {
    mockFilms = films;
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      json: () => Promise.resolve(mockFilms)
    }));
  })

  it('should match the snapshot', () => {
    const wrapper = shallow(<App />);

    expect(wrapper).toMatchSnapshot();
  })

  it('Should render with the correct default state', () => {
    const wrapper = shallow(<App />);

    expect(wrapper.state('currentRandomFilm')).toEqual({});
    expect(wrapper.state('currentSectionData')).toEqual([]);
  });

  it('fetchfilm return an array of films', async () => {
    const wrapper = shallow(<App />)
    const expected = mockFilms
    const getRandomFilm = jest.fn().mockImplementation(() => {
      return expected
    });
    const actual = await wrapper.instance().fetchFilm()

    expect(actual).toEqual(expected);
  })

  it('handle favorite should send a favorite', () => {
    const wrapper = shallow(<App />);
    const mockCurrentCards = [{
      homeworld: "Tatooine",
      homeworldPopulation: "200000",
      name: "Luke Skywalker",
      species: "Human"
    },
    {
      homeworld: "Naboo",
      homeworldPopulation: "4500000000",
      name: "R2-D2",
      species: "Droid"
    }
  ]
  // wrapper.state('currentSectionData') = mockCurrentCards

    console.log(wrapper.state())
    const mockEvent = {target: {value: 'Luke Skywalker'}}
  })
})
