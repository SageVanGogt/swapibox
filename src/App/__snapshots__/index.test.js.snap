// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <nav>
          <Button
                    handleClickEvent={[Function]}
                    section="people"
          />
          <Button
                    handleClickEvent={[Function]}
                    section="planets"
          />
          <Button
                    handleClickEvent={[Function]}
                    section="vehicles"
          />
</nav>,
        <Film
          currentRandomFilm={Object {}}
/>,
        <CardContainer
          currentSectionData={Array []}
/>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Button
              handleClickEvent={[Function]}
              section="people"
/>,
            <Button
              handleClickEvent={[Function]}
              section="planets"
/>,
            <Button
              handleClickEvent={[Function]}
              section="vehicles"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleClickEvent": [Function],
              "section": "people",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleClickEvent": [Function],
              "section": "planets",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleClickEvent": [Function],
              "section": "vehicles",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "nav",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "currentRandomFilm": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "currentSectionData": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <nav>
            <Button
                        handleClickEvent={[Function]}
                        section="people"
            />
            <Button
                        handleClickEvent={[Function]}
                        section="planets"
            />
            <Button
                        handleClickEvent={[Function]}
                        section="vehicles"
            />
</nav>,
          <Film
            currentRandomFilm={Object {}}
/>,
          <CardContainer
            currentSectionData={Array []}
/>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Button
                handleClickEvent={[Function]}
                section="people"
/>,
              <Button
                handleClickEvent={[Function]}
                section="planets"
/>,
              <Button
                handleClickEvent={[Function]}
                section="vehicles"
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleClickEvent": [Function],
                "section": "people",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleClickEvent": [Function],
                "section": "planets",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleClickEvent": [Function],
                "section": "vehicles",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "nav",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "currentRandomFilm": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "currentSectionData": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
